File: tools/bench/dp_window_max.cpp
// VIII-10 의 수치 — 창 최댓값을 다시 훑는 DP 대 단조 덱 DP (g++ 13, -O2).
//
//   g++ -std=c++17 -O2 -o /tmp/dpwm tools/bench/dp_window_max.cpp && /tmp/dpwm
//
// 점화식은 본문과 같다.  dp[i] = a[i] + max(dp[j]) for j in [i-k, i-1]
#include <algorithm>
#include <chrono>
#include <cstdio>
#include <deque>
#include <random>
#include <vector>
using namespace std;

long long naive(const vector<int>& a, int k) {
    int n = (int)a.size();
    vector<long long> dp(n);
    dp[0] = a[0];
    for (int i = 1; i < n; i++) {
        long long best = dp[i - 1];
        int lo = max(0, i - k);
        for (int j = lo; j < i; j++) best = max(best, dp[j]);
        dp[i] = a[i] + best;
    }
    return dp[n - 1];
}

long long fast_dq(const vector<int>& a, int k) {
    int n = (int)a.size();
    vector<long long> dp(n);
    dp[0] = a[0];
    deque<int> dq;
    for (int i = 1; i < n; i++) {
        while (!dq.empty() && dp[dq.back()] <= dp[i - 1]) dq.pop_back();
        dq.push_back(i - 1);
        while (dq.front() < i - k) dq.pop_front();
        dp[i] = a[i] + dp[dq.front()];
    }
    return dp[n - 1];
}

template <class F>
double timed(F f, const vector<int>& a, int k, int reps, long long& out) {
    vector<double> ts;
    for (int r = 0; r < reps; r++) {
        auto t0 = chrono::steady_clock::now();
        out = f(a, k);
        ts.push_back(chrono::duration<double>(chrono::steady_clock::now() - t0).count());
    }
    sort(ts.begin(), ts.end());
    return ts[ts.size() / 2];
}

int main() {
    mt19937 rng(11);
    uniform_int_distribution<int> small(-9, 9);

    // 정확성 — 작은 입력에서 두 구현을 대조
    for (int t = 0; t < 2000; t++) {
        int n = 2 + (int)(rng() % 8), k = 1 + (int)(rng() % 4);
        vector<int> a(n);
        for (int& x : a) x = small(rng);
        if (naive(a, k) != fast_dq(a, k)) { printf("불일치\n"); return 1; }
    }
    printf("accuracy: naive == fast on 2000 random cases\n\n");

    uniform_int_distribution<int> big(-1000, 1000);
    printf("        n |      k | naive(sec) | fast(sec) | ratio | same\n");
    printf("----------|--------|------------|-----------|-------|-----\n");
    for (auto [n, k] : vector<pair<int, int>>{{200000, 2000}, {500000, 2000}, {1000000, 2000}}) {
        vector<int> a(n);
        for (int& x : a) x = big(rng);
        long long v1 = 0, v2 = 0;
        double t1 = timed(naive, a, k, 1, v1);
        double t2 = timed(fast_dq, a, k, 3, v2);
        printf("%9d | %6d | %10.3f | %9.4f | %5.0f | %s\n", n, k, t1, t2, t1 / t2,
               v1 == v2 ? "yes" : "NO");
    }
    return 0;
}

File: tools/bench/dp_window_max.py
#!/usr/bin/env python3
"""VIII-10 의 수치 — 창 최댓값을 다시 훑는 DP 대 단조 덱 DP (CPython 3.13).

    python3.13 tools/bench/dp_window_max.py

점화식은 본문과 같다.
    dp[0] = a[0]
    dp[i] = a[i] + max(dp[j])   for j in [i-k, i-1]

같은 답이 나오는지 먼저 확인하고, 그다음 시간을 잰다.
Python 은 O(nk) 쪽이 커지면 측정이 끝나지 않으므로 n 을 작게 잡는다 —
환산값은 적지 않는다. 자릿수는 잰 구간에서 이미 갈린다.
"""
from __future__ import annotations

import random
import statistics
import sys
import time
from collections import deque


def naive(a: list[int], k: int) -> int:
    n = len(a)
    dp = [0] * n
    dp[0] = a[0]
    for i in range(1, n):
        best = dp[i - 1]
        lo = i - k
        if lo < 0:
            lo = 0
        for j in range(lo, i):
            if dp[j] > best:
                best = dp[j]
        dp[i] = a[i] + best
    return dp[-1]


def fast(a: list[int], k: int) -> int:
    n = len(a)
    dp = [0] * n
    dp[0] = a[0]
    dq: deque[int] = deque()
    for i in range(1, n):
        while dq and dp[dq[-1]] <= dp[i - 1]:
            dq.pop()
        dq.append(i - 1)
        while dq[0] < i - k:
            dq.popleft()
        dp[i] = a[i] + dp[dq[0]]
    return dp[-1]


def timed(fn, a, k, reps=3):
    ts = []
    out = None
    for _ in range(reps):
        t0 = time.perf_counter()
        out = fn(a, k)
        ts.append(time.perf_counter() - t0)
    return out, statistics.median(ts)


def main() -> int:
    rnd = random.Random(11)

    # 1) 정확성 — 작은 입력 2,000개를 브루트포스와 대조
    for _ in range(2000):
        n = rnd.randint(2, 9)
        k = rnd.randint(1, 4)
        a = [rnd.randint(-9, 9) for _ in range(n)]
        if naive(a, k) != fast(a, k):
            print("불일치:", a, k)
            return 1
    print("정확성: 무작위 2,000건에서 naive == fast")

    print()
    print("      n |      k | naive(초) | fast(초) | 배수 | 답 일치")
    print("--------|--------|-----------|----------|------|--------")
    for n, k in ((20_000, 2_000), (50_000, 2_000), (100_000, 2_000)):
        a = [rnd.randint(-1000, 1000) for _ in range(n)]
        v1, t1 = timed(naive, a, k, reps=1)
        v2, t2 = timed(fast, a, k, reps=3)
        print(f"{n:7,} | {k:6,} | {t1:9.3f} | {t2:8.3f} | {t1 / t2:4.0f} | {v1 == v2}")
    return 0


if __name__ == "__main__":
    sys.exit(main())
